implement flush-timeout() for http

bazsi opened the pull request:
This branch implements the flush-timeout() option for LogThreadedDestDriver, which works very similarly to what we had earlier (e.g. pre syslog-ng 3.3).

Then it makes this option available to the HTTP destination driver. This should fix CPU consumption issues for elasticsearch, such as #2131.

After giving some though whether we should simply use the old option name (e.g. flush-timeout()) or introduce a new one, I've finally concluded on simply using the old one, meaning that it should be brought out of deprecation. But looking at it closely, that deprecation wasn't very successful, in the sense it has not been documented.

This also means that the flush-timeout() and flush-lines() options would be inherited automatically from the global options. This branch implements that as well for the generic case, and then disables this inheritance in the http driver, as that would be unexpected.

NOTE: this is not a complete solution for #2131. To solve #2131, we would need to add this option to the Java binding as well, or feed Elastic with the http destination. The latter should be an interesting experiment.

furiel commented on lib/logthrdestdrv.c:
```diff
@@ -303,9 +294,9 @@ _perform_flush(LogThreadedDestDriver *self)
+  if (!self->suspended)
     {
-      worker_insert_result_t result = self->worker.flush(self);
+      worker_insert_result_t result = log_threaded_dest_worker_flush(self);
```
This change is not an equivalent transformation. In the original version, accepting the batch happens only if flush vfunction is provided. In the new code it happens everytime. This can only be a problem when a driver wants to use the batching framework without specifying a flush. I cannot tell any reasonable usecase for this,  though the API allows to do so.

The cleanest would be to start the function with:
```
if (!self->worker.flush)
  return;
```

lbudai replied:
@furiel: I think it is equivalent. Just check the implementation of `log_threaded_dest_worker_flush`.
(and IMHO it is better as checking of the existence of a virtual method is moved to the right place)

furiel replied:
I still think it is not equivalent, at least not mathematically. In the original implementation with self->worker.flush == 0, the whole if block is skipped. In the new implementation, we simulate a dummy flush method by returning `WORKER_INSERT_RESULT_SUCCESS`. But that has a side effect: acknowledging a half-full batch, which the original implementation did not do. Maybe we can reason why it is not a problem (I am not 100% convinced yet), but this is a functional change in the middle of an otherwise innocent-looking refactor patch, which is unfortunate.

bazsi replied:
I agree this is not functionally equivalent, however after giving some thought this should have been the way it was coded originally. I should have added more comments to the commit message, sorry for that.

As for the rationale:

Remember that in earlier versions of this code, on_message_queue_empty() was called if provided and flush superceedes that logic. 

The original on_message_queue_empty() method was returning void, basically assuming that any flush that needed happening indeed happened successfully, there was no way to indicate failure.

With the new flush logic, in this codepath, one of two things could happen:

*1) the flush() method is provided*

In this case  we work as expected.

*2) flush() is not set*

then I'd expect the driver never return a QUEUED verdict as there's no way to handle queued messages. So
instead of QUEUED it would return a final verdict like SUCCESS or ERROR. In either case the insert() method would already call _process_result(), causing batch_size to be reset to zero. This in turn means that the process_result() here would become a noop, as anything it does is basically dependant on batch_size. (well, it also resets the retries_counter, but again that shouldn't matter that we set that to zero twice).

furiel replied:
Thanks for the detailed explanation. This was the missing puzzle for me, without flush but returning QUEUED users cannot accept batch. I see `_accept_batch` is not exposed to the `logthrdestdrv.h` header. (Not static function either, maybe worth change it to static later.).

furiel commented on lib/logthrdestdrv.c:
```diff
@@ -325,16 +363,40 @@ _message_became_available_callback(gpointer user_data)
+  if (self->suspended)
+    _schedule_restart_on_suspend_timeout(self);
+  else if (!_flush_time_has_elapsed(self))
+    _schedule_restart_on_flush_timeout(self);
```
Here in the `!_flush_time_has_elapsed(self)` branch we should still call `iv_task_register(&self->do_work);`, if there are messages in the queue, we should handle them.

lbudai replied:
@furiel: flush_handler and work_handler is the same : `_perform_work`, so I think it is correct, but maybe I'm wrong.

furiel replied:
They almost do the same, but `flush_handler` executes `_perform_work` only `flush_timeout` later. But when we are in the `!suspended` branch, we should have a message to be handled. It is possible that is the last message missing from a batch to flush, which would delay the batch flush unnecessarily.

bazsi replied:
do_work would wake up immediately and not when the message arrives. This means that we would spin in the thread, as we would wake up immediately, finding that the queue is empty.

The way this is handled is that log_queue_check_items() registers a callback function whenever there is NO message in the queue and that gets called if one is added, this function is _message_became_available_callback() and that wakes up our thread immediately.

So in summary:
1) this function is only called when we have some items in the batch but no items in the queue
2) in this case message_became_available_callback() is registered and would be called by the LogQueue
3) any new message thus wakes up the thread, which would pull all of the messages it finds there.
4) if the batch is still not complete we would repeat the same process until either the flush timeout elapses, or we get enough items  to fill a batch.

I think this should be what we expect. I agree though that the wakeup logic is a bit messy here. I tried a few alternatives, but none seemed to read better :(

lbudai commented:
@bazsi : `test_logthrdestdrv`  is broken currently

bazsi commented:
The last set of changes broke the test (e.g. that one that migrates this to be flush-timeout and implements inheritance from the global options), but works for me now.

bazsi commented:
I've also added two testcases to test_logthrdestdrv to test flush-timeout(), there are some timing magic involved so please review it with a critical eye.

furiel commented:
@bazsi I do not have any other comments. The unit test for flush timeout looks good to me. When travis and kira is green, we can merge this.

bazsi commented:
@kira-syslogng retest this please

bazsi commented:
This is now ready for merging, unfortunately I've rebased it a number of times invalidating the reviews so far.

The important changes were:
1) fix a behavioural change, flush-timeout() defaults to 10 seconds for destinations that do not override it specifically (e.g. java). But that shouldn't be applied unless at least one message is QUEUED into the batch.

2) in the course of fixing 1) above, I've introduced a race condition that I've also fixed, but took me a few iterations until I've found the root cause. I've squashed the issue and its fix into the base patch, so it is not visible with the current set of patches.

3) a few improvements in .travis.yml on macOS, as that platform was the most reliable in finding the race mentioned in 2) above.

furiel commented on lib/logthrdestdrv.c:
```diff
@@ -243,6 +274,7 @@ _process_result(LogThreadedDestDriver *self, gint result)
       break;
 
     case WORKER_INSERT_RESULT_QUEUED:
+      self->enable_flush_timeout = TRUE;
```
Should we set this variable to false in _accept_batch and _drop_batch?

furiel replied:
Oh ok I thought we could use self->flush_timeout, but that would not work because it can be set with global options, so we need an additional filtering mechanism not to start the timer for drivers that do not support it. Marking as resolved.

bazsi commented:
No, it is only a "turn-on" mechanism, to be compatible with legacy drivers,
which shouldn't be subject to flush-timeout() at all.

On Mon, Sep 10, 2018 at 10:31 AM, furiel <notifications@github.com> wrote:

> *@furiel* commented on this pull request.
> ------------------------------
>
> In lib/logthrdestdrv.c
> <https://github.com/balabit/syslog-ng/pull/2251#discussion_r216236535>:
>
> > @@ -243,6 +274,7 @@ _process_result(LogThreadedDestDriver *self, gint result)
>        break;
>
>      case WORKER_INSERT_RESULT_QUEUED:
> +      self->enable_flush_timeout = TRUE;
>
> Should we set this variable to false in _accept_batch and _drop_batch?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/2251#pullrequestreview-153670020>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AArldjg2RlLf7iIZtOBZCMiNf3mDQHx7ks5uZiNFgaJpZM4WOV_d>
> .
>

bazsi commented:
Thanks \o/
